filesetup.c

```diff
@@ -338,6 +338,95 @@ static bool pre_read_file(struct thread_data *td, struct fio_file *f)
 	return ret;
 }
 
+/*
+ * Prepopulate regular file with data.
+ * Leaves f->fd open on success, caller must close.
+ */
+static int prepopulate_file(struct thread_data *td, struct fio_file *f)
+{
+	int flags;
+	unsigned long long left, bs;
+	char *b = NULL;
+
+	if (read_only) {
+		log_err("fio: refusing extend of file due to read-only\n");
+		return 0;
+	}
+	if (td->o.fill_device) {
+		log_err("fio: prepopulate option with fill_device makes no sense\n");
+		return 0;
+	}
+
+	flags = O_WRONLY;
+	if (td->o.allow_create)
+		flags |= O_CREAT;
+
+#ifdef WIN32
+	flags |= _O_BINARY;
+#endif
+
+	dprint(FD_FILE, "open file %s, flags %x\n", f->file_name, flags);
+	f->fd = open(f->file_name, flags, 0644);
+	if (f->fd < 0) {
+		int err = errno;
+
+		if (err == ENOENT && !td->o.allow_create)
+			log_err("fio: file creation disallowed by "
+					"allow_file_create=0\n");
+		else
+			td_verror(td, err, "open");
+		return 1;
+	}
+
+	left = f->real_file_size; // XXX ?? f->file_offset + f->io_size
```

Comment:
and need a little help here - which one is correct? or maybe should cover both (if io_size is set)

Reply:
Now I believe I should use file_off + io_size, but I'm not sure how to apply offset into the `while` loop with writes.

(I moved this issue into a new PR)